street_network/src/transform/dual_carriageways.rs

```diff
@@ -1,55 +1,350 @@
+                    .or_insert_with(Vec::new)
+                    .push(r);
+            }
+        }
+
+        // Look for a group of 3. Two should be one-way for driving, the other shouldn't.
+        for (road_name, groups) in roads_by_name {
+            if groups.len() != 3 {
+                continue;
+            }
+            let mut oneway_roads = Vec::new();
+            let mut bidi_roads = Vec::new();
+            for r in groups {
+                if streets.roads[&r].oneway_for_driving().is_some() {
+                    oneway_roads.push(r);
+                } else {
+                    bidi_roads.push(r);
+                }
+            }
+
+            if oneway_roads.len() != 2 || bidi_roads.len() != 1 {
+                continue;
+            }
+
+            // Preserving notes about old detection:
+            // - look for dual_carriageway=yes tag, but since it's rarely there, maybe just use as an "opt
+            //   in" technique
+            // - make sure the 2 oneway roads are within 10 degrees of the bidi road?
+            // - The two one-ways should point at each other
+            // - Maybe one intersection could be a MultiConnection for two roads? For now just take one.
+
+            return Some(Self {
+                i,
+                side1: oneway_roads.remove(0),
+                side2: oneway_roads.remove(0),
+                road_name,
+            });
+        }
+
+        None
+    }
+
+    fn debug(&self, streets: &StreetNetwork) {
+        streets.debug_intersection(self.i, "join/split that isnt DC");
+        streets.debug_road(self.side1, "side1 of failed DC");
+        streets.debug_road(self.side2, "side2 of failed DC");
+    }
+}
+
+// Step 2: trace both sequences of one-ways between the intersections where a dual carriageway
+// splits/rejoins
+struct DualCarriagewayPt1 {
+    road_name: String,
+    i1: osm::NodeID,
+    i2: osm::NodeID,
+    // side1 points from i1 to i2
+    side1: Vec<OriginalRoad>,
+    // side2 points from i2 to i1
+    side2: Vec<OriginalRoad>,
+}
+
+impl DualCarriagewayPt1 {
+    fn new(streets: &StreetNetwork, mc: &MultiConnection) -> Option<Self> {
+        let (side1, i2_v1) = Self::trace_side(streets, &mc.side1, mc.i, &mc.road_name)?;
+        let (side2, i2_v2) = Self::trace_side(streets, &mc.side2, mc.i, &mc.road_name)?;
+
+        // TODO Something very odd has happened. Make a new copy of the map for debugging and label
+        // the strangeness.
+        if i2_v1 != i2_v2 {
+            return None;
+        }
+
+        let mut side1 = Self::orient_oneways(side1)?;
+        let mut side2 = Self::orient_oneways(side2)?;
+
+        // Which one goes from i1->i2?
+        let i1 = mc.i;
+        let i2 = i2_v1;
+        for swap in [false, true] {
+            if swap {
+                std::mem::swap(&mut side1, &mut side2);
+            }
+            let side1_endpts = (side1[0].i1, side1.last().as_ref().unwrap().i2);
```

Comment:
Maybe we need a `struct` to represent a list of roads in order, with methods to walk the intersections in order or get endpoints easily